velox/common/base/BloomFilter.h

```diff
@@ -42,18 +59,65 @@ class BloomFilter {
     bits_.resize(std::max<int32_t>(4, bits::nextPowerOfTwo(capacity) / 4));
   }
 
+  bool isSet() {
+    return bits_.size() > 0;
+  }
+
   // Adds 'value'.
-  void insert(uint64_t value) {
+  // case:
+  // InputType can be one of folly hasher support type when hashInput = false.
+  // InputType can only be uint64_t when hashInput = true.
+  template <class InputType = uint64_t, bool hashInput = true>
+  void insert(InputType value) {
     set(bits_.data(),
         bits_.size(),
-        hashInput ? folly::hasher<uint64_t>()(value) : value);
+        hashInput ? folly::hasher<InputType>()(value) : value);
   }
 
-  bool mayContain(uint64_t value) const {
+  // case:
+  // InputType can be one of folly hasher support type when hashInput = false.
+  // InputType can only be uint64_t when hashInput = true.
+  template <class InputType = uint64_t, bool hashInput = true>
+  bool mayContain(InputType value) const {
     return test(
         bits_.data(),
         bits_.size(),
-        hashInput ? folly::hasher<uint64_t>()(value) : value);
+        hashInput ? folly::hasher<InputType>()(value) : value);
+  }
+
+  void merge(const char* serialized) {
+    auto stream = initializeInputStream(serialized);
+    auto size = stream.read<int32_t>();
+    bits_.resize(size);
+    auto bitsdata =
+        reinterpret_cast<const uint64_t*>(serialized + stream.offset());
+    if (bits_.size() == 0) {
+      for (auto i = 0; i < size; i++) {
+        bits_[i] = bitsdata[i];
+      }
+      return;
+    } else if (size == 0) {
+      return;
+    }
+    VELOX_DCHECK_EQ(bits_.size(), size);
+    for (auto i = 0; i < size; i++) {
+      bits_[i] |= bitsdata[i];
```

Comment:
If I change to
```
    bits::orBits(bits_.data(), bitsdata, 0, size);
    // for (auto i = 0; i < size; i++) {
    //   bits_[i] |= bitsdata[i];
    // }
```
@Yuhta  This test will failed:
```
/mnt/DP_disk1/code/velox/build/velox/common/base/tests# ./velox_base_test --gtest_filter=*merge*
Note: Google Test filter = *merge*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from BloomFilterTest
[ RUN      ] BloomFilterTest.merge
/mnt/DP_disk1/code/velox/velox/common/base/tests/BloomFilterTest.cpp:90: Failure
Value of: bloom.mayContain<int32_t>(i)
  Actual: false
Expected: true
/mnt/DP_disk1/code/velox/velox/common/base/tests/BloomFilterTest.cpp:90: Failure
Value of: bloom.mayContain<int32_t>(i)
  Actual: false
Expected: true
/mnt/DP_disk1/code/velox/velox/common/base/tests/BloomFilterTest.cpp:90: Failure
Value of: bloom.mayContain<int32_t>(i)
  Actual: false
Expected: true
/mnt/DP_disk1/code/velox/velox/common/base/tests/BloomFilterTest.cpp:90: Failure
Value of: bloom.mayContain<int32_t>(i)
  Actual: false
Expected: true
/mnt/DP_disk1/code/velox/velox/common/base/tests/BloomFilterTest.cpp:90: Failure
Value of: bloom.mayContain<int32_t>(i)
  Actual: false
Expected: true
/mnt/DP_disk1/code/velox/velox/common/base/tests/BloomFilterTest.cpp:90: Failure
Value of: bloom.mayContain<int32_t>(i)
  Actual: false
Expected: true
/mnt/DP_disk1/code/velox/velox/common/base/tests/BloomFilterTest.cpp:90: Failure
Value of: bloom.mayContain<int32_t>(i)
  Actual: false
Expected: true
/mnt/DP_disk1/code/velox/velox/common/base/tests/BloomFilterTest.cpp:90: Failure
Value of: bloom.mayContain<int32_t>(i)
  Actual: false
Expected: true
/mnt/DP_disk1/code/velox/velox/common/base/tests/BloomFilterTest.cpp:90: Failure
Value of: bloom.mayContain<int32_t>(i)
  Actual: false
Expected: true
/mnt/DP_disk1/code/velox/velox/common/base/tests/BloomFilterTest.cpp:90: Failure
Value of: bloom.mayContain<int32_t>(i)
  Actual: false
Expected: true
[  FAILED  ] BloomFilterTest.merge (0 ms)
[----------] 1 test from BloomFilterTest (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] BloomFilterTest.merge

```

Reply:
`size` here is number of words, you need to pass number of bits.  Do it like this:

```
bits::orBits(bits_.data(), bitsdata, 0, 64 * size);
```

Reply:
Thanks! That worked.